### PR TITLE
Extend download repository functionality, Fixes #811

### LINKED ISSUE #811: Extend Download Repo Functionality
At the moment, instructors and tutors can download multiple student solutions (in form of git repos) for the manual correction in the Exercise Scores page (e.g. https://artemis.ase.in.tum.de/#/course/29/exercise/438/dashboard) of programming exercises, see the following screenshot after clicking on `Download Repos`:

image

After clicking on `Download Repositories`, the server clones the repos of all specified students (or pulls them if they already have been cloned, e.g. for the online editor), executes some additional steps (see e.g. the paper Stager: Simplifying the Manual Assessment of Programming Exercises
 on http://ceur-ws.org/Vol-2358/paper-03.pdf), creates a zip file for each repository and then creates a zip file for all repos. This is done in `ExerciseService.java` in the method `exportParticipations(...)`.

This functionality should be extended:
1) Allow instructors to choose to download all student repos of this exercise instead of manually specifying some students (this functionality should be used in the future for downloading all repos to execute plagiarism checks and here the instructors typically want to download all repos)
2) Normalize code style: Unify the encoding and line endings of all files in the repository to UTF-8 and unix line endings, see TODO in `ExerciseService.java` in line 355. This might be tricky and was not implemented yet, but if you google, there should be a lot of options how to accomplish this.
3) Filter late submissions, see TODOs in `GitService.java in line 380ff: basically the commit hash of the last submission within the exercise due date from the Artemis database should be used here to checkout a specific version of the repository --> Additionally allow the user to specify a custom date instead of the exercise due date in the dialog on the client side (see below)
4) Clone the repositories into a different folder on the Artemis server and NOT into the one that is already used by Artemis. Otherwise, this might interfere with the usage of the students in the online editor.

The functionality should also be customizable in the future:
1) Allow the user to choose in the client dialog which of the tasks should be executed with checkboxes and boolean parameters in the corresponding REST call:
    a) Filter late submissions with custom date (default is exercise due date)
    b) Add student to project name
    c) Squash after instructor
    d) Normalize code style

In addition, the functionality should be moved into the `ProgrammingExerciseService`

### PR DESCRIPTION
**Pull request by sleiss**
### Checklist
- [x] I tested the changes and all related features on the test server https://artemistest.ase.in.tum.de.
- [x] I documented my source code using the JavaDoc / JSDoc style.
- [x] I added integration test cases for the server (Spring) related to the features
- [x] ~I added integration test cases for the client (Jest) related to the features~
- [x] I added screenshots/screencast of my UI changes
- [x] I translated all the newly inserted strings

### Motivation and Context
This fixes #811: It adds several options when downloading the student repos.
Also, now a different folder for the repos is used in order to prevent inference with the code students use.

### Description
- The repositories of all students can now be downloaded by activating the checkbox. This disables the text-box with the student ids (and the student ids won't be evaluated on the server, even if something is written in there)
- Filtering of late submissions can now be activated/disabled. By default, the exercise due date is used as date, but it can be changed manually. Also, the commithash saved within the submission is now used instead of the commithash of the git repository itself if possible in order to avoid students from creating commits with a wrong date and tricking the system with that. If this is not possible, the old solution is used as fallback.
- It can now be toggled if the student names should be added to the project or not.
- It can now be toggled if all commits made by the student should be squashed into one commit.
- The code style can now be normalized: This changes the encoding of all files to UTF-8 (the apache tika library is used in order to determine the encoding) and changes all line endings to UNIX line endings.

### Steps for Testing

1. Log in to Artemis
2. Navigate to Course Administration
3. Click on exercises for one course.
4. Click on 'Scores' for a programming exercise
5. Click on 'Download repos'
6. Test if the available checkboxes work correctly
7. Test if the newly added folder ('repos-download' by default) is used instead of the folder used by students ('repos' by default)

### Screenshots
(somehow the mouse is missing..)
downloadRepo

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/ExerciseService.java [hunk 1]
```diff
@@ -338,39 +379,47 @@ public void cleanup(Long id, boolean deleteRepositories) {
         for (StudentParticipation participation : exercise.getStudentParticipations()) {
             ProgrammingExerciseStudentParticipation studentParticipation = (ProgrammingExerciseStudentParticipation) participation;
             try {
-                if (studentParticipation.getRepositoryUrl() == null || studentParticipation.getStudent() == null
-                        || !studentIds.contains(studentParticipation.getStudent().getLogin())) {
+                if (!allStudents && (studentParticipation.getRepositoryUrl() == null || studentParticipation.getStudent() == null
```
Comment (thilo-behnke) [inline src/main/java/de/tum/in/www1/artemis/service/ExerciseService.java]: I don't like the design of this whole method, I'll pick the argument `allStudents` vs `studentIds` here as an example and explain what we could improve:
- Instead of having two arguments, we could just pass in the concerned participations as a list. Then the method knows which repositories to export and only has to iterate over them
- A method that calls this method can then only load the participations from the db that are needed for the export operation

The rest of the parameters (e.g. `filterLateSubmissions`) could be grouped into a new Class called `RepositoryExportOptions`. This way the method has less parameters and we can document the options better.

File: src/main/java/de/tum/in/www1/artemis/service/ExerciseService.java [hunk 2]
```diff
@@ -380,7 +429,8 @@ public void cleanup(Long id, boolean deleteRepositories) {
         try {
             // create a large zip file with all zipped repos and provide it for download
             log.debug("Create zip file for all repositories");
-            zipFilePath = Paths.get(zippedRepoFiles.get(0).getParent().toString(), exercise.getCourse().getTitle() + " " + exercise.getTitle() + studentIds.hashCode() + ".zip");
+            zipFilePath = Paths.get(zippedRepoFiles.get(0).getParent().toString(),
+                    exercise.getCourse().getTitle() + " " + exercise.getTitle() + (studentIds != null ? studentIds.hashCode() : "ALL") + ".zip");
```
Comment (thilo-behnke) [inline src/main/java/de/tum/in/www1/artemis/service/ExerciseService.java]: I wonder if we have to make this change, I would just use e.g. the timestamp as a hash here.

File: src/main/java/de/tum/in/www1/artemis/service/FileService.java [hunk 3]
```diff
@@ -503,4 +504,116 @@ public void replaceVariablesInFile(String filePath, List<String> targetStrings,
...
+     */
+    public void normalizeLineEndingsRecursive(String startPath) throws IOException {
+        log.debug("Normalizing file endings in directory {}", startPath);
+        File directory = new File(startPath);
+        if (!directory.exists() || !directory.isDirectory()) {
+            throw new RuntimeException("File endings in directory " + startPath + " should be normalized but the directory does not exist.");
+        }
+
+        // Get all files in directory
+        String[] files = directory.list((current, name) -> new File(current, name).isFile());
```
Comment (thilo-behnke) [inline src/main/java/de/tum/in/www1/artemis/service/FileService.java]: This method can return null under certain circumstances. I would propose using propose using FileUtils here: https://commons.apache.org/proper/commons-io/javadocs/api-2.5/org/apache/commons/io/FileUtils.html#listFiles(java.io.File,%20org.apache.commons.io.filefilter.IOFileFilter,%20org.apache.commons.io.filefilter.IOFileFilter)

Comment (thilo-behnke) [reply src/main/java/de/tum/in/www1/artemis/service/FileService.java]: Then you would also get the recursive look-up for free.

File: src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java [hunk 4]
```diff
@@ -82,14 +85,29 @@ public GitService() {
...
      */
     public Repository getOrCheckoutRepository(ProgrammingExerciseParticipation participation) throws IOException, InterruptedException, GitAPIException {
+        return getOrCheckoutRepository(participation, REPO_CLONE_PATH);
+    }
+
+    /**
+     * Get the local repository for a given participation. If the local repo does not exist yet, it will be checked out.
+     *
+     * @param participation Participation the remote repository belongs to.
+     * @param repoClonePath path where the repo is located on disk
```
Comment (thilo-behnke) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java]: I would propose `targetPath` for this argument.

File: src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java [hunk 5]
```diff
@@ -366,31 +398,47 @@ public ObjectId getLastCommitHash(URL repoUrl) throws EntityNotFoundException {
...
+    @Transactional(readOnly = true)
+    public void filterLateSubmissions(Repository repository, ProgrammingExerciseStudentParticipation participation, ZonedDateTime filterLateSubmissionsDate) {
+        if (filterLateSubmissionsDate == null) {
+            // No date set in client and exercise has no due date
             return;
         }
 
         try {
             Git git = new Git(repository);
+            Optional<Submission> lastValidSubmission = participation.getSubmissions().stream().filter(s -> s.getSubmissionDate().isBefore(filterLateSubmissionsDate))
```
Comment (thilo-behnke) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java]:
```suggestion
            Optional<Submission> lastValidSubmission = participation.getSubmissions().stream().filter(s -> s.getSubmissionDate().isBefore(filterLateSubmissionsDate)).max(Comparator.comparing(Submission::getSubmissionDate));
```

This is a recommendation from Intellij.

File: src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java [hunk 6]
```diff
@@ -366,31 +398,47 @@ public ObjectId getLastCommitHash(URL repoUrl) throws EntityNotFoundException {
...
      * @param repository Local Repository Object.
-     * @param exercise   ProgrammingExercise associated with this repo.
+     * @param participation   ProgrammingExerciseStudentParticipation associated with this repo.
+     * @param filterLateSubmissionsDate the date after which all submissions should be filtered out (may be null)
      */
-    public void filterLateSubmissions(Repository repository, ProgrammingExercise exercise) {
-        if (exercise.getDueDate() == null) {
-            // No dates set on exercise
+    @Transactional(readOnly = true)
+    public void filterLateSubmissions(Repository repository, ProgrammingExerciseStudentParticipation participation, ZonedDateTime filterLateSubmissionsDate) {
```
Comment (thilo-behnke) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java]: I think it would be better if the GitService would just be provided the `validSubmission` instead of providing all submissions and a date to filter.
This logic is unrelated to the GitService's responsibilities.

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java [hunk 7]
```diff
@@ -312,29 +315,54 @@ private StatsForInstructorDashboardDTO populateCommonStatistics(Exercise exercis
...
         // - one boolean flag per stager task (see exportParticipations)
         // - one boolean flag that all student submissions should be downloaded
 
+        ZonedDateTime filterLateSubmissionsZonedDateTime;
+
+        if (filterLateSubmissionsDate == null) {
+            filterLateSubmissionsZonedDateTime = exercise.getDueDate();
+        }
+        else {
+            filterLateSubmissionsZonedDateTime = ZonedDateTime.ofInstant(filterLateSubmissionsDate.toInstant(), ZoneId.systemDefault());
```
Comment (thilo-behnke) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java]: Isn't this just `ZonedDateTime.now()`?

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java [hunk 8]
```diff
@@ -312,29 +315,54 @@ private StatsForInstructorDashboardDTO populateCommonStatistics(Exercise exercis
...
+     * @param squashAfterInstructor whether all changes after the setup from the instructors should be squashed into one commit
+     * @param normalizeCodeStyle whether the code style should be normalized (line endings, encoding)
      * @return ResponseEntity with status
      * @throws IOException if submissions can't be zipped
      */
     @GetMapping(value = "/exercises/{exerciseId}/participations/{studentIds}")
     @PreAuthorize("hasAnyRole('TA', 'INSTRUCTOR', 'ADMIN')")
-    public ResponseEntity<Resource> exportSubmissions(@PathVariable Long exerciseId, @PathVariable String studentIds) throws IOException {
-        studentIds = studentIds.replaceAll(" ", "");
+    public ResponseEntity<Resource> exportSubmissions(@PathVariable Long exerciseId, @PathVariable String studentIds, @RequestParam boolean allStudents,
```
Comment (thilo-behnke) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java]: As mentioned above, we could group the options into an object.

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java [hunk 9]
```diff
@@ -312,29 +315,54 @@ private StatsForInstructorDashboardDTO populateCommonStatistics(Exercise exercis
...
-        List<String> studentList = Arrays.asList(studentIds.split("\\s*,\\s*"));
-        if (studentList.isEmpty()) {
-            return ResponseEntity.status(HttpStatus.BAD_REQUEST).headers(HeaderUtil.createAlert(applicationName, "Given studentlist for export was empty or malformed", ""))
-                    .build();
+        File zipFile;
+        if (allStudents) {
+            zipFile = exerciseService.exportParticipationsAllStudents(exerciseId, filterLateSubmissions, filterLateSubmissionsZonedDateTime, addStudentName, squashAfterInstructor,
+                    normalizeCodeStyle);
         }
+        else {
```
Comment (thilo-behnke) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java]: As mentioned above, we could load the necessary participations by the studentIds and then call the same method for zipping their repos.

File: src/main/webapp/app/scores/exercise-scores-repo-export-dialog.component.ts [hunk 10]
```diff
@@ -28,27 +42,38 @@ export class ExerciseScoresRepoExportComponent {
...
-            err => {
-                this.exportInProgress = false;
-            },
-        );
+        const studentIdList = this.studentIdList !== undefined && this.studentIdList !== '' ? this.studentIdList.split(',').map(e => e.trim()) : ['ALL'];
+        this.exerciseService
+            .exportRepos(
+                exerciseId,
+                studentIdList,
+                this.allStudents !== undefined ? this.allStudents : false,
```
Comment (thilo-behnke) [inline src/main/webapp/app/scores/exercise-scores-repo-export-dialog.component.ts]: Again, if you would use an object for the settings, you would have less parameters in the class and you could easily set defaults.

File: src/main/webapp/app/scores/exercise-scores-repo-export-dialog.component.ts [hunk 11]
```diff
@@ -28,27 +42,38 @@ export class ExerciseScoresRepoExportComponent {
...
+                this.normalizeCodeStyle !== undefined ? this.normalizeCodeStyle : true,
+            )
+            .subscribe(
+                response => {
+                    this.jhiAlertService.success('Export of repos was successful. The exported zip file with all repositories is currently being downloaded');
+                    this.activeModal.dismiss(true);
+                    this.exportInProgress = false;
+                    if (response.body) {
+                        const zipFile = new Blob([response.body], { type: 'application/zip' });
+                        const url = window.URL.createObjectURL(zipFile);
```
Comment (thilo-behnke) [inline src/main/webapp/app/scores/exercise-scores-repo-export-dialog.component.ts]: You can import window as a dependency in the component's constructor (which is preferred).

File: src/main/java/de/tum/in/www1/artemis/service/ExerciseService.java (hunk 2 above)
Comment (sleiss) [reply src/main/java/de/tum/in/www1/artemis/service/ExerciseService.java]: In the current setup, we have to make this change, as the studentIds list can be null (therefore calling .hashCode() can produce a NullPointerException), but I agree that using a timestamp is better suited here.

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java (hunk 7 above)
Comment (sleiss) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java]: I don't think so:
`filterLateSubmissionsDate` is the date the user selected in the client and as it is not null, the user actually selected a specific date.
As this can (and most likely will) differ from the current date, I believe it's correct the way it is.

File: src/main/java/de/tum/in/www1/artemis/service/FileService.java [hunk 12]
```diff
@@ -503,4 +507,104 @@ public void replaceVariablesInFile(String filePath, List<String> targetStrings,
...
     }
+
+    /**
+     * This normalizes all line endings to UNIX-line-endings recursively from the startPath.
+     *
+     * @see {@link #normalizeLineEndings(String) normalizeLineEndings}
+     * @param startPath          the path where the start directory is located
+     * @throws IOException if an issue occurs on file access for the normalizing of the line endings.
+     */
+    public void normalizeLineEndingsRecursive(String startPath) throws IOException {
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/service/FileService.java]: I think the name is misleading, as this implementation does not utilize recursion.

File: src/main/java/de/tum/in/www1/artemis/service/FileService.java [hunk 13]
```diff
@@ -503,4 +507,104 @@ public void replaceVariablesInFile(String filePath, List<String> targetStrings,
...
+    }
+
+    /**
+     * This converts all files to the UTF-8 encoding recursively from the startPath.
+     *
+     * @see {@link #convertToUTF8(String) convertToUTF8}
+     * @param startPath          the path where the start directory is located
+     * @throws IOException if an issue occurs on file access when converting to UTF-8.
+     */
+    public void convertToUTF8Recursive(String startPath) throws IOException {
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/service/FileService.java]: I think the name is misleading, as this implementation does not utilize recursion.

File: src/main/java/de/tum/in/www1/artemis/service/FileService.java (hunk 12 above)
Comment (sleiss) [reply src/main/java/de/tum/in/www1/artemis/service/FileService.java]: This was a remnant of an earlier implementation, I changed it to be more specific.

File: src/main/java/de/tum/in/www1/artemis/service/FileService.java (hunk 13 above)
Comment (sleiss) [reply src/main/java/de/tum/in/www1/artemis/service/FileService.java]: This was a remnant of an earlier implementation, I changed it to be more specific.

**Comment by krusche** [comment]
> Complicated stuff, good work!
> In general it works, I have a couple of comments about the design of methods, etc. though (see below).
> 
> Some comments about the functionality:
> 
> 1. Minor: The zip folder structure is nested when exporting a single repository. Also the single student repositories within the main zip are all themselves zipped. This might be due to file size, but I thought it was annoying when testing the feature.
> 
> image
> 
> 1. Medium: There is no autocomplete for the student names. This would help the instructors/tutors to efficiently export repositories.
> 2. Major: When I enter one student login that does not exist, nothing happens after clicking on download repos
> 3. Major: No progress is visualized for the user, even though the download could take a while (let's say exporting 2000 repositories). We currently have a 15 min limit a request can take on nginx, this might even take longer.
>    A simple solution would be to at least show a loading indicator in the UI while the request has not returned a response (we have a new generic button component for this `jhi-button` that already has a loading state built in).
>    A more sophisticated solution would involve more work and is out-of-scope for this PR, but running an async task that sends websocket messages to the client, which are then visualized with e.g. a progress bar would be idea.
> 
> On a side note: There are no server tests in this PR which I think is critical for methods with heavy logic (e.g. normalizeFileEndings or squashingCommits). We should write unit tests for those.

Thanks, these are all good suggestions for a follow up issue / PR. Right now, we want to give this functionality to the users. We can still improve its usability later on.
